target/linux/ramips/files/drivers/net/ethernet/ralink/gsw_mt7620.c

```diff
@@ -97,9 +97,12 @@ static void mt7620_hw_init(struct mt7620_gsw *gsw, int mdio_mode)
 	mtk_switch_w32(gsw, mtk_switch_r32(gsw, GSW_REG_MIB_CNT_EN) | (1 << 1), GSW_REG_MIB_CNT_EN);
 
 	if (mdio_mode) {
-		/* turn off ephy and set phy base addr to 12 */
+		/* turn off ephy and set phy base address to 0b01000 unless specified */
+		if (!gsw->ephy_base)
+			gsw->ephy_base = 8;
```

Comment:
this one can be assigned at line 282 where the OF property is read:
```c
	if (of_property_read_u16(np, "mediatek,ephy-base-address", &val) == 0)
		gsw->ephy_base = val;
	else
		gsw->ephy_base = 0;
```

Reply:
if I put it here, I think I would have to call the function that returns mdio_mode, because otherwise, it is only available in the function `mt7620_hw_init`

so we would have
`elif (mt7620_mdio_mode(priv->dev->of_node))`

is that correct or am I missing something?

Reply:
I think the best way to do this is get rid of `mt7620_mdio_mode` function and move the lines somewhere else...

working on that now

Reply:
Oops. I misread the code when writing that comment.

I think having two different default values doesn't make sense here. This is a board-specific property and should be overridden with dts when needed. Default value are supposed to be used in most cases and the value you are putting here  is just for that single ar8327/qca8337 board so it should be set in dts.
(Not every switch expose their PHYs directly on host mdio. e.g. mt7530 doesn't do that.)

Reply:
BTW I'm not asking a revert of that refactor to mt7620_mdio_mode. Its still great to separate property reading from actual hw init :)

Reply:
After adding the 2 boards I have and after @polex adds his board, setting ephy_base to 8 would be needed for a total of 5 boards, and there could be more in the future.

I think it is safe for the driver to handle this, it would not negatively affect other switches that don't need it

The only case I can imagine where you need a custom value, is when the PHY is at an MDIO address > 7. I think this is very rare, since there currently is no mt7620 device like that.

Also worth mentioning again, I got the default value of 8 from reading the registers in uboot for my board.


> Not every switch expose their PHYs directly on host mdio. e.g. mt7530 doesn't do that

would it be better, rather, that for these boards with mt7530 that you remove mdio-mode property? Maybe there is something I don't understand yet, because to me this means the switch doesn't use MDIO at all...

Reply:
being a little more clear:

mt7530 boards have defined PHYs 0, 1, 2, 3, 4, and 1f

this default value moves internal mt7620 PHYs to 8, 9, 10, 11, 12

is there a reason 0-4 is defined with mt7530?

Reply:
@981213 see commit 194ca6127ee18cd3a95da4d03f02e43b5428c0bb

"extra init that we did not do when using an external mt7530"

all of those lines added is signaled in the driver with mdio_mode

It looks like mdio_mode is not needed at all for boards with mt7530

Reply:
more reference 5a6229a93df8a3f292094012a43410c6a24c210e

after reading this one I understand better

Reply:
After reading both commits you should know that 194ca61 uses "mediatek,mt7530" to say "this board is using external mt7530" and it's actually writing registers via mdio to the external mt7530 chip. I would say 5a6229a is a crazy commit because "mdio-mode" in no way indicates that an external mt7530 is used.
Similarly I'm going against this default value because mdio-mode does not indicate "I want to move internal PHY address to somewhere else". That's what an explicit "mediatek,ephy-base-address" is supposed to do.

Reply:
After reading both commits you should know that 194ca61 uses "mediatek,mt7530" to say "this board is using external mt7530" and it's actually writing registers via mdio to the external mt7530 chip. I would say 5a6229a is a crazy commit because "mdio-mode" in no way indicates that an external mt7530 is used.
Similarly I'm going against this default value because mdio-mode does not indicate "I want to move internal PHY address to somewhere else". That's what an explicit "mediatek,ephy-base-address" is supposed to do.

Reply:
wow, so it seems that me and many others have misread the code here, and that led to unnecessary commits...

with this explanation, the comment and function names in that block makes sense now

So, do we rename mdio_mode to something like has_mt7530, or do we move those lines to the other driver?